### ISSUE TITLE
[HOLD for payment 2023-03-21] [$1000] [Feature] Create autosuggestion UI for :emoji_codes:

### ISSUE THREAD

**Issue by roryabraham** [Expensify/App#12188]
Let's add an auto-suggestion UI for emojis so that they can be added inline in messages. Right now, typing `:smile:` works, but only if you know that it works. Let's make it easier to add emojis to messages using an inline emoji selector. 

![image](https://user-images.githubusercontent.com/1371996/211650171-8aa634fa-255e-4f52-bec7-3a39ebf51c44.png)

It will work like this: 
1. As soon as you type colon and two more characters without a space (e.g. `:sm`) we start searching (using the [suggestEmojis function](https://github.com/Expensify/App/blob/c4937290090396262084edc453eb6bb9b499d01a/src/libs/EmojiUtils.js#L221))
2. We display the matches in the auto-suggestion modal, in the order returned by the function, with the closest match at the top
a. We display the top suggestion highlighted. 
b. Pressing enter or tab selects the highlighted option. 
c. The arrow keys allow you to cycle through the options. The list should “wrap”, such that if you have the first element in the list highlighted and you press ArrowUp, then you are taken to the last element to the list. Similarly, if you have the last element in the list highlighted and you press ArrowDown, then you are taken back up to the first element in the list
d. Every time your search query updates (i.e: you type another letter), then the first item is refocused. The reason for this is that when you update the search query we should assume that the “best result” is the one displayed first in the list, so that’s the one you’re most likely to want
4. If there are no matches, we don't display the auto-suggestion modal
5. If the user types a space or a colon (e.g. `:smi ` or `:smi:`) we stop searching and don't show the auto-suggestion modal
6. If they tap, click, press enter, or press tab on one of the emoji options then we finish the text (e.g. `:smi` becomes `:smile: `) and add a space so that the user can continue typing 
a. When we finish the text, we should just replace it with the actual emoji `code` from emoji.js (which is returned by the search method) instead of the `:<name>:` since that would then need to be converted again

**Comment by roryabraham** [comment]
@stephanieelliott I think we should probably start this issue at $1000

**Comment by stephanieelliott** [comment]
Thanks @roryabraham! 

Posted the job to Upwork :https://www.upwork.com/jobs/~01a345ace891dbce3d

**Comment by getusha** [comment]
Should the emoji popup appear after we typed **:emojicode:** or while we typing e.g **:emo...** and suggestions will start to appear
What do you think?

**Comment by shawnborton** [comment]
Yup, happy to take a fresh look here. Did we ever have mobile mocks for this one too?

**Comment by stitesExpensify** [comment]
> Yup, happy to take a fresh look here. Did we ever have mobile mocks for this one too?

Not that I could find, no.

**Comment by getusha** [comment]
isn't that great if the suggested emojies aligned horizontally?

**Comment by afmire877** [comment]
What is there a list of emojis that should be available, as there could be literally infinite?

I found this list => https://unicode.org/emoji/charts/full-emoji-list.html but as a user, I would not expect some of these.

**Comment by stitesExpensify** [comment]
This is our list of emojis https://github.com/Expensify/App/blob/main/assets/emojis.js

There is already a trie data structure that will get suggestions [here](https://github.com/Expensify/App/blob/main/src/libs/EmojiUtils.js#L229-L254). This issue is mostly just UI and keybinds utilizing the existing code

**Comment by getusha** [comment]
@stitesExpensify @shawnborton  @mananjadhav 
Take a look at this.

https://user-images.githubusercontent.com/75031127/198814481-bb39fd2b-01ee-4f8d-9088-c9c46f14a2d1.mp4

What do you think about this. Kindly share your idea and I will submit the proposal.

**Comment by mananjadhav** [comment]
I like the compact view of the emojis. I am wondering we should be showing the `:emojicode:` adjacent to the emoji?

Also can you tell me how does this UI look when I have my composer expanded? If you type few lines of text, you'll see an expand composer icon.

**Comment by roryabraham** [comment]
Lets please pause on any further proposals on this issue until @shawnborton gets the opportunity to provide some updated mockups. Thanks.

**Comment by shawnborton** [comment]
I brought up some [conversation in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1667265243767019) but I actually quite like the horizontal idea that @getusha proposed above. Let's see if the group agrees with that approach first and then I can create some clean mocks.

**Comment by mananjadhav** [comment]
@shawnborton It looks like we've got good +1 (and 0 Nos) for the horizontal approach. Can we have a quick design for this one?

I am guessing it should also include the placement of the popup when the composer is in the full screen mode?

**Comment by shawnborton** [comment]
Here is what I am thinking:

![image](https://user-images.githubusercontent.com/2319350/199564332-6ce668ae-d5c7-4c9a-bdd8-fc491b61d512.png)

**Comment by shawnborton** [comment]
For the full screen composer, maybe we need to move it to the bottom of the screen?

**Comment by shawnborton** [comment]
Actually let's hold on that, still some active discussion

**Comment by shawnborton** [comment]
Here is a version with the vertical list:
![image](https://user-images.githubusercontent.com/2319350/199592132-697d7235-a8df-4269-ba17-4e8a931d58c1.png)


Still waiting for some convo to shake out.

**Comment by Karim-30** [comment]
### Proposal

The biggest challenge in this feature is on large screen devices the suggestion menu should appear above the writing cursor, but in React Native there is nothing can tell us where is the writing cursor on the x-axis.

 we solved it by creating a hidden `<Text>` (paddingBottom: -166) has the same style that composer has and the same text, at the end of its content we are putting a blank `<Text>` and give it a `ref`, now we can get the cursor location using this `ref` it is equal `offsetLeft` of the blank `<Text>`.

On IOS, there is no `offsetLeft` property and I think it is very hard to get the cursor position, so the suggestion menu should take the full screen width on IPad  similar to iPhone.

 Also `onSelectionChange` Event on IOS provide a poor data, for example if the text inside the composer is `:` then we pressed `s` we should see `:s` in the event but we only see `:` this causes a delay (suggestion menu appears after the 2nd char).

I haven’t test on Android yet, I hope `onSelectionChange`  Event provide a good data.

At the current time the suggestion menu only appears when sending a comment not when editing.

I know the code looks bad and not easy to understand but it is a prototyping and we are improving it.



**Code**

in `src/styles/styles.js` : 

```diff
    emojiHeaderStyle: {
        backgroundColor: themeColors.componentBG,
        width: '100%',
        ...spacing.pv3,
        fontFamily: fontFamily.GTA_BOLD,
        fontWeight: fontWeightBold,
        color: themeColors.heading,
        fontSize: variables.fontSizeSmall,
    },

+   emojiSuggestionsContainer: {
+       backgroundColor: colors.white,
+       borderRadius: 12,
+       borderWidth: 1,
+       borderColor: themeColors.border,
+       justifyContent: 'center',
+       overflow: 'hidden',
+       boxShadow: variables.popoverMenuShadow,
+       position: 'absolute',
+       bottom: 66,
+   },

+   emojiSuggestionsText: {
+       fontFamily: fontFamily.EMOJI_TEXT_FONT,
+       fontSize: variables.fontSizeMedium,
+       marginRight: 20,
+   },

```

in `src/styles/variables.js` : 
```diff
    sliderBarHeight: 8,
    sliderKnobSize: 26,
+   emojiSuggestionsWidth: 280,
```

Copy `EmojiSuggestions` folder in the attached to `src/components` folder

 Replace `src/pages/home/report/ReportActionCompose.js` content with the one in the attached.



**Result** 


https://user-images.githubusercontent.com/108357004/199657578-2f5c075c-0daa-4d9b-8bac-cd1cc8fb16eb.mov


https://user-images.githubusercontent.com/108357004/199657610-e9ffc7c6-ca8c-42a2-8694-bccbff916a78.mov


[autosuggestion.zip](https://github.com/Expensify/App/files/9926110/autosuggestion.zip)

**Comment by shawnborton** [comment]
That's looking pretty good @Karim-30 ! Yeah, it seems like we should go with the vertical version you have there. I am cool with the desktop version following the cursor, but can you also share what we could do for the full screen compose view? I think in that context we might need to also push the menu below the cursor depending on where the cursor is in the window. 

Also, [Figma file is here](https://www.figma.com/file/LsaunhlRi8SjGEz7pPkyk0/Emojis?node-id=0%3A1) if you want to take a closer look at the mocks.

**Comment by JmillsExpensify** [comment]
>I think in that context we might need to also push the menu below the cursor depending on where the cursor is in the window.

Agreed. I bet it's below the cursor in the vast majority of cases, but to your point, it has to work in both places depending on where the cursor is at.

**Comment by Karim-30** [comment]
>I think in that context we might need to also push the menu below the cursor depending on where the cursor is in the window.

we can do it using `offsetTop` property, it represent the location of the `<Text>` we are using as a `ref` on the y-axis.

**Comment by mananjadhav** [comment]
@shawnborton Can we confirm here if vertical UI is what we are going ahead with?

**Comment by shawnborton** [comment]
Confirmed, let’s do vertical

**Comment by mananjadhav** [comment]
Okay thanks for the confirmation @shawnborton. 

I can see @Karim-30's proposal aligned with the requirements. But I would wait for a day to see if @getusha is keen on submitting a revised proposal for vertical design?

**Comment by getusha** [comment]
Thanks @mananjadhav
I will submit the revised proposal in short.

**Comment by mananjadhav** [comment]
@getusha do you have any revised proposal. Please reply at the earliest to avoid any confusion later on. I want to be fair with @Karim-30 too, who's already submitted the proposal. Will wait for a day.

**Comment by getusha** [comment]
You're right @mananjadhav 
It took me unexpectedly longer to add the features.
You can review @Karim-30's proposal. Thanks.

**Comment by roryabraham** [comment]
In case it wasn't clear from @shawnborton's comments above, this issue is 🟢 for proposals again.

@mananjadhav let us know your thoughts on @Karim-30's proposal. I think it looks overall pretty good, but had some questions for @Karim-30:

> we solved it by creating a hidden <Text> (paddingBottom: -166) has the same style that composer has and the same text, at the end of its content we are putting a blank <Text> and give it a ref

Clever solution, but did you investigate adding the blank `<Text>` component as a child of the `<TextInput>` in `<Composer>`? As you can see [here](https://github.com/facebook/react-native/blob/114098d419e2a7a254f81b9eb9f0cc542194a660/Libraries/Components/TextInput/TextInput.js#L1450-L1458), it's possible to use `<TextInput>` as a wrapper component, and the children to the `<TextInput>` get wrapped in a (nestable) `<Text>` component. So maybe we could make put the blank `<Text>` component directly into the `<Composer>`'s `<TextInput>` and not need to have the extra hidden component?

Also, whenever we use hidden components, I think we should use `opacity: 0` to hide them instead of just adjusting their position to move them off-screen, because that will be more predictable/reliable

Maybe a more robust solution would be to figure out how to implement a React Native feature request that emits a [`TextLayoutEvent`](https://reactnative.dev/docs/text#textlayoutevent) for the selected text along with the [`SelectionChangeEvent`](https://github.com/facebook/react-native/blob/b966d297245a4c1e2c744cfe571396cfa7e5ffd3/Libraries/Components/TextInput/TextInput.js#L107-L112). 🤔 I tried to look into how that might work, and couldn't figure it out (yet, anyways)... I'm guessing we'd need to find a way to execute [`TextLayoutManager->measureLines`](https://github.com/facebook/react-native/blob/3fb3ce4fa2080cdb7737184100040c15201484d4/ReactCommon/react/renderer/textlayoutmanager/platform/cxx/TextLayoutManager.h#L41-L48) from [`RCTBaseTextInputView::textInputDidChangeSelection`](https://github.com/facebook/react-native/blob/96027f787ade1a465569975059426d2c03ee5ad0/Libraries/Text/TextInput/RCTBaseTextInputView.m#L504-L518) (iOS) and [`ReactInputManager::onSelectionChanged`](https://github.com/facebook/react-native/blob/b51a350c1ea679c9edb8d102eca299a2cb04db14/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputManager.java#L1204-L1222) (Android)

**Comment by roryabraham** [comment]
Also, one thing I'll mention for people leaving proposals on this issue as well as @mananjadhav – we need the emoji list that pops up to be navigable via the `ArrowUp` and `ArrowDown` keys. You should utilize the [`ArrowKeyFocusManager`](https://github.com/Expensify/App/blob/4eae63c7e62b2cca48f0c306869840876b2f14e6/src/components/ArrowKeyFocusManager.js) component to do that.

**Comment by getusha** [comment]
as @shawnborton's design mockups there is no emoji_code highlighting, is that intentional?
also, making the EmojiSuggestion horizontal when the composer get's expanded might be good idea?

**Comment by roryabraham** [comment]
> On IOS, there is no offsetLeft property and I think it is very hard to get the cursor position

@Karim-30 What about using a [`LayoutEvent`](https://reactnative.dev/docs/layoutevent) for the empty Text, and another `LayoutEvent` for the parent component (i.e: a wrapper around the `TextInput` in `Composer`, for example), and then taking the difference between the two to compute `offsetX` and `offsetY` in a platform-agnostic way?

**Comment by Karim-30** [comment]
>maybe we could make put the blank <Text> component directly into the <Composer>'s <TextInput> and not need to have the extra hidden component?

@roryabraham This will work If the writing cursor is at the end of the text but won’t if it is in the middle.

i.e: Lets assume that our hidden component looks like this :

```
<Text name=“same_as_compose”>{this.state.hiddenText}<Text name=“blank_text” ref={} /></Text>
```

Now if the text inside the composer is `Fewfefwef : fdsfdsfds` and the cursor is next to `wef :` in this case, and depending on what is happening in `onSelectionChange` method the `hiddenText` should be `Fewfefwef :` and the blank `<Text>`  should be next to it so the menu will appear above the cursor, not above the last char in the composer.

![Screen Shot 2022-11-10 at 11 15 27 AM](https://user-images.githubusercontent.com/108357004/201123208-30937519-54ac-4f84-9b4f-8891e54dbcda.png)




> What about using a [LayoutEvent](https://reactnative.dev/docs/layoutevent) for the empty Text, and another LayoutEvent for the parent component ?

I think we have a logical solution now by using `onTextLayout` for `same_as_composer <Text>`, the advantage of `onTextLayout` over `onLayout` is that `onTextLayonut` got called every time `this.state.hiddenText` is changed, not related to the text width and height, you will notice the difference if you have a multiline text.

**The solution will be:**

- When the text inside the compose changes we get its value in `onChangeText`Event, this value is the same as what is written in the composer(doesn’t have a one char delay as in `onSelectionChange` Event on IOS)

- If the user changes the cursor position, in `onSelectionChange` Event we have the composer value(we got it in step 1) and the cursor position(`e.nativeEvent.selection.end`)  now we can get `this.state.hiddenText`.

- When `this.state.hiddenText` changes, `onTextLayout` got called, and it returns the number of lines and each line’s width and height, so if it returns one line this means `left: width_of_this_line, bottom: 66` , and if it returns multiple lines this means `left: last_line_width, bottom: 66 + sum_of_lines_height`

This solution logically should work on all platforms, let's see if it is applicable or not.

**Comment by Karim-30** [comment]
>I can say there's good amount of refactor required in terms of the component, especially blocks in `onSelectionChange` and `render`

I totally agree.

>I hope you've gone through [@roryabraham](https://github.com/roryabraham)'s comments on using ArrowKeyFocusManager ?

@mananjadhav The `ArrowUp`, `ArrowDown`, `Enter` and `Escape` all work in [this proposal](https://github.com/Expensify/App/issues/12188#issuecomment-1301705641), as you can see I selected the `:eyes:` emoji using the keyboard in the first video, we only added some logic in [triggerHotkeyActions](https://github.com/Expensify/App/blob/1e30fb51097392bfce6c45a9b635714b238e1449/src/pages/home/report/ReportActionCompose.js#L417-L442) method to make these work.


I tried the solution I mentioned [here](https://github.com/Expensify/App/issues/12188#issuecomment-1310426957), it worked fine on iPad and Tablet.

https://user-images.githubusercontent.com/108357004/201348353-5c6191a6-70e1-4d50-a5bb-44473a046d8a.mov


https://user-images.githubusercontent.com/108357004/201348394-b3c6b486-063e-4fa4-b9f4-c9b55af32b4d.mov




 The problem is `onTextLayout` doesn’t work on Web or Desktop, so we can use the old approach (`Ref` + `offsetLeft`) for Web and Desktop, and `onTextLayout` for IOS and Android.

**Comment by mananjadhav** [comment]
Waiting on @roryabraham to share their feedback.

**Comment by shawnborton** [comment]
Not overdue

**Comment by roryabraham** [comment]
Okay, I think we talked about it a bit in slack [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1667983099982879). Overall/long-term I think a better solution would be an implementation involving substantial changes in the React Native upstream `<TextInput>` component. Something like this:

- Update `TextInput` such that it gives us an `onTextLayout` prop that, like the `onTextLayout` prop for `Text`, gets called with a [`TextLayoutEvent`](https://reactnative.dev/docs/text#textlayoutevent).
- Update `TextInput`’s `onSelectionChange` prop such that it provides not only the new selection range but also a `TextLayoutEvent` for the selected/highlighted portion of the text
- Then we can use the `TextLayoutEvent` provided by `onSelectionChange` to locate the cursor/selection in the text input and position the emoji picker UI relative to that

However, we all agreed that implementing this would be much much harder, and without having those new features in React Native, I think @Karim-30's solution would be a great working solution we could do in a shorter timeline.

So I think we need to choose:

1. **The harder but cleaner way:** upgrade React Native to provide the layout information we want it to
1. **The more hacky way that would be easier and we know would work:** Implement @Karim-30's clever workaround with a couple invisible text components.

I think I would vote for **The harder but cleaner way** rather than using a workaround.

Another approach would be to do the workaround that @Karim-30 suggested first so we get this feature out more quickly, then create a follow-up issue to implement **The harder but cleaner way**. Definitely want some other opinions on the best way to go here.

**Comment by puneetlath** [comment]
@roryabraham @stitesExpensify To help us clear out the large backlog of /App bugs, we're putting the spotlight every bug in the repo already than 4 weeks old. To help unblock the roadmap and get our bug pipeline back in equilibrium, can you:  
 
- Decide whether any proposals currently meet our guidelines and can be approved as-is 
- For any that can't, please take this issue internal and treat it as one of your highest priorities 
- If you have any questions, don't hesitate to start a discussion in #bug-zero

**Comment by stephanieelliott** [comment]
1 vote here for the **harder but cleaner way** -- IMO it really doesn't really make sense to implement a stopgap solution to get this in product quicker. I would feel differently if were something that significantly impacted functionality or improved user experience, but this feels _very_ "nice to have" and not something we need to solve for ASAP

**Comment by mananjadhav** [comment]
I am not sure about the urgency of the issue, but yeah if there isn't a constraint on the urgency, then 👍  for **harder but cleaner way**

**Comment by puneetlath** [comment]
I agree with the harder, but cleaner way. But I also think we should probably punt this until after WAQ is done. cc @JmillsExpensify

**Comment by puneetlath** [comment]
I agree with the harder, but cleaner way. But I also think we should probably punt this until after WAQ is done. cc @JmillsExpensify

**Comment by JmillsExpensify** [comment]
@puneetlath Yeah agreed. This is filed as post-WAQ polish in the tracking issue, though I'll add a HOLD to this issue as well for clarity.

**Comment by JmillsExpensify** [comment]
Ah in fact, I'm the BZ assigned on the emoji tracking issue and I should have assigned myself at that point, but I forgot to on this one. I'll go ahead and do that now, and also update this issue to $1000 for when we get back to it.

**Comment by stitesExpensify** [comment]
Moving this to weekly so we don't get pinged as much while it's on hold

**Comment by JmillsExpensify** [comment]
I'd even go with monthly for now.

**Comment by JmillsExpensify** [comment]
Removed the Help Wanted label since we're on hold.

**Comment by JmillsExpensify** [comment]
This is nice polish to pick up again once WAQ is behind us.

**Comment by JmillsExpensify** [comment]
Officially removing the HOLD. This would be an ideal one to consider picking up.

**Comment by shawnborton** [comment]
I think we confirmed that we want to do the vertical style:
![image](https://user-images.githubusercontent.com/2319350/210748931-3016c6e8-cb71-451b-bbb6-2b457492aa54.png)

Just need to update that mock for dark mode.

**Comment by JmillsExpensify** [comment]
@puneetlath is going to pick this one up and work on this with Margelo, so I'm going to assign him to this issue as well.

**Comment by puneetlath** [comment]
Just to confirm the way this should work: 

1. As soon as you type colon and one more character (e.g. `:s`) we start searching (using the same search method used by the emoji picker)
2. We display up to 5 matches in the auto-suggestion UI, in the same order the would appear in the emoji picker, with the closest match at the top
a. On platforms with a physical keyboard, we display the top suggestion highlighted. Pressing enter selects it. The arrow keys allow you to cycle through the options. 
4. If there are no matches, we don't display the auto-suggestion UI
5. If the user types a space or a colon (e.g. `:smi ` or `:smi:`) we stop searching and don't show the auto-suggestion UI
6. If they tap, click, or press enter on one of the emoji options then we finish the text (e.g. `:smi` becomes `:smile: `) and add a space so that the user can continue typing 

Questions: 
1. Why only 5 matches? If it's just for height, we could keep the height at 5 while still allowing to scroll to more options. 
2. Should pressing escape close the auto-suggestion UI? It does in Slack and Discord for example. 
3. How should the modal be aligned with the text on wider screens? In the mockup it is left-aligned with the compose box. Should that always be the case or should it be aligned with where I type? Here are two examples. Discord has it be the full width of the compose box. Whereas slack has it follow the text. 

<img width="728" alt="image" src="https://user-images.githubusercontent.com/1371996/211095633-a74f2bdc-ab71-4172-b5b1-8fb2d6c2e8a9.png">

<img width="661" alt="image" src="https://user-images.githubusercontent.com/1371996/211095758-1dfc3cd1-e5b7-462e-be8f-33a923ffb414.png">

**Comment by stitesExpensify** [comment]
Confirmations:
1: 
> we start searching (using the same search method used by the emoji picker)

It actually uses a different search method because we created a `trie` data structure that searches faster. There is an issue [here](https://github.com/Expensify/Expensify/issues/239099) to move the picker search to that method as well. 

2-4 sound good to me

5: 
> If they tap, click, or press enter on one of the emoji options then we finish the text (e.g. :smi becomes :smile: ) and add a space so that the user can continue typing

When they click we should just replace it with the actual emoji `code` from emoji.js (which is returned by the search method) instead of the `:<name>:` since that would then need to be converted again. I also think pressing `tab` should select the emoji 

Questions:
1. No reason really, it's just what github does. We can change it to however many we want though (Original explanation [here](https://github.com/Expensify/App/pull/10165#discussion_r967149090))

2. Makes sense to me! Sounds like a great QOL addition

3. I will leave that to @shawnborton 😄

**Comment by JmillsExpensify** [comment]
Agreed, great questions. 

> Why only 5 matches? If it's just for height, we could keep the height at 5 while still allowing to scroll to more options.

I like that idea, and if we go that route then for clarity, pressing up/down on the keyboard should navigate accordingly through your search list. 

> Should pressing escape close the auto-suggestion UI? It does in Slack and Discord for example.

Agreed.

> How should the modal be aligned with the text on wider screens? In the mockup it is left-aligned with the compose box. Should that always be the case or should it be aligned with where I type? Here are two examples. Discord has it be the full width of the compose box. Whereas slack has it follow the text.

I don't think there is a right answer here, but I prefer Slack's approach. I'd imagine we'd do the same thing for mentions as well, align the auto-complete menu to where you type.

**Comment by shawnborton** [comment]
Yeah, Slack's approach is nice as it lets you keep your eyes focused on the current spot in which you are typing.

**Comment by shawnborton** [comment]
Let me know if we need updated mocks for this, or if that's something Margelo will handle.

**Comment by puneetlath** [comment]
@shawnborton a mock would be great for this. Maybe mobile and desktop.

**Comment by shawnborton** [comment]
Let me know if these work:
![image](https://user-images.githubusercontent.com/2319350/211532273-2dbb8cf8-7818-4824-894a-551eed891f98.png)

**Comment by roryabraham** [comment]
I'd like to clarify one aspect of the expected behavior here.

1. When the auto-suggestion UI appears, the first element is highlighted by default
1. Then using the ArrowUp and ArrowDown keys should move the highlight up and down through the list
1. The list should “wrap”, such that if you have the first element in the list highlighted and you press ArrowUp, then you are taken to the last element to the list. Similarly, if you have the last element in the list highlighted and you press ArrowDown, then you are taken back up to the first element in the list
1. Every time your search query updates (i.e: you start typing), then the first item is refocused. The reason for this is that when you update the search query we should assume that the “best result” is the one displayed first in the list, so that’s the one you’re most likely to want
1. Pressing Enter or Tab with an item highlighted replaces what you’ve got so far with the emoji and closes the auto-suggestion UI

**Comment by puneetlath** [comment]
Updated the OP based on the discussion.

**Comment by JmillsExpensify** [comment]
Those look great.

**Comment by mananjadhav** [comment]
@JmillsExpensify @puneetlath Is this going to be handled internally? 

Should we also highlight the popup positioning? We had some discussions around it [here](https://github.com/Expensify/App/issues/12188#issuecomment-1308384898) , [here](https://github.com/Expensify/App/issues/12188#issuecomment-1310821579) and [here](https://github.com/Expensify/App/issues/12188#issuecomment-1317776123).

**Comment by stitesExpensify** [comment]
The plan is to take this internally and work with Margelo on the solution. If anything changes I'll be sure to let you know!

**Comment by mananjadhav** [comment]
Thanks for the update @stitesExpensify

**Comment by shawnborton** [comment]
Not overdue

**Comment by roryabraham** [comment]
Latest discussion [here](https://expensify.slack.com/archives/C035J5C9FAP/p1673893200304789)

**Comment by JmillsExpensify** [comment]
Is this currently taking a backseat to emoji reactions? @hannojg I'd personally take this one over mobile emoji search.

**Comment by hannojg** [comment]
Yeah, we are currently prioritising emoji reactions (if that's correct with you guys?). We are currently onboarding a new member at Margelo and he will pick up this and the emoji search task in the coming days!

**Comment by Szymon20000** [comment]
Do you already know how the popup will track cursor position?
I saw hard and clean solution above and the smart one that will work right away. We can also try to find a different solution.
But first want to know if that was resolved somewhere already.

**Comment by perunt** [comment]
@roryabraham could you assign me to this issue?

**Comment by JmillsExpensify** [comment]
Done! Excited to have you help out on this one. 🙌🏼

**Comment by roryabraham** [comment]
> Do you already know how the popup will track cursor position?

Unless there is some problem with it, I would guess we would want to use our existing [Popover](https://github.com/Expensify/App/tree/main/src/components/Popover) or [PopoverWithMeasuredContent](https://github.com/Expensify/App/blob/66a05aa4015f96f5bab59929698d517a616620d1/src/components/PopoverWithMeasuredContent.js) components.

**Comment by Szymon20000** [comment]
Sure we can use it. I actually asked about the second part of the problem. I guess Popover component let us display any component we want to based on coordinates that we pass to it. However, what's problematic a bit is how we want to get coordinates of a cursor. I saw your comment here https://github.com/Expensify/App/issues/12188#issuecomment-1308384898. And was wondering if you already know how you want it to be implemented or should we find our own solution.

**Comment by shawnborton** [comment]
Not overdue

**Comment by stitesExpensify** [comment]
Hi @Szymon20000 you can feel free to find your own solution! We've had some ideas, but have not made a definitive decision.

**Comment by shawnborton** [comment]
Not overdue

**Comment by stitesExpensify** [comment]
Hi @Szymon20000 @perunt is there any update here?

**Comment by perunt** [comment]
Attached PR. Going to record a video of how it works so you can decide if you like this approach

**Comment by stitesExpensify** [comment]
Great, looking forward to the video!

**Comment by perunt** [comment]
https://user-images.githubusercontent.com/48593211/215896031-7980f6bd-fc42-4025-96c1-0867d126df27.mp4

@stitesExpensify the rest videos are in the Screenshots/Videos section in the PR

**Comment by stitesExpensify** [comment]
Here is the web version with the static auto-suggestion (as opposed to following the cursor). Any thoughts @roryabraham @JmillsExpensify @puneetlath @shawnborton ? 

https://user-images.githubusercontent.com/48593211/215886393-7b1e2914-c264-4f49-8539-f3f0b8e9dbbe.mp4

**Comment by puneetlath** [comment]
Looks great to me!

**Comment by JmillsExpensify** [comment]
Love it! This is great.

**Comment by stitesExpensify** [comment]
Looks like Shawn is OOO, so let's move forward with the static version! @perunt LMK when the PR is ready for review!

**Comment by mananjadhav** [comment]
Quick bump here as @shawnborton is back I guess.

**Comment by shawnborton** [comment]
Works for me too!

**Comment by JmillsExpensify** [comment]
Confirming that everything is unblocked now?

**Comment by stitesExpensify** [comment]
Yep! The PR is being reviewed

**Comment by MelvinBot** [comment]
@JmillsExpensify, @shawnborton, @mananjadhav, @stitesExpensify, @roryabraham, @perunt Whoops! This issue is 2 days overdue. Let's get this updated quick!

**Comment by JmillsExpensify** [comment]
Still working through the linked PR.

**Comment by MelvinBot** [comment]
@JmillsExpensify, @shawnborton, @mananjadhav, @stitesExpensify, @roryabraham, @perunt Eep! 4 days overdue now. Issues have feelings too...

**Comment by MelvinBot** [comment]
@JmillsExpensify, @shawnborton, @mananjadhav, @stitesExpensify, @roryabraham, @perunt Still overdue 6 days?! Let's take care of this!

**Comment by stitesExpensify** [comment]
This PR has been heavily reviewed, and my best guess is that it should be
live in the next 7-10 days

**Comment by stitesExpensify** [comment]
2 approvals, just waiting for a final C+ review and we'll be good to go!

**Comment by JmillsExpensify** [comment]
Let's get this one over the finish line this week! Can we resolve conflicts in the PR, make the requested updates, and merge today?

**Comment by stitesExpensify** [comment]
Yep it was just updated again last night and should be merged today!

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.83-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/14686


If no regressions arise, payment will be issued on 2023-03-21. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by roryabraham** [comment]
@JmillsExpensify bump for C+ payment

**Comment by mananjadhav** [comment]
Accepted @JmillsExpensify

**Comment by JmillsExpensify** [comment]
Awesome, all paid out! We should be all done here, closing the issue.
